File: src/TinRoll.Data/Repository/UserRepository.cs
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using TinRoll.Data.Entities;
using TinRoll.Data.Repository.Interface;

namespace TinRoll.Data.Repository
{
    public class UserRepository : IUserRepository
    {
        private IRepository<User> _baseRepo;

        public UserRepository(IRepository<User> baseRepo)
        {
            _baseRepo = baseRepo;
        }

        public async Task<User> CreateUserAsync(User user)
        {
            return await _baseRepo.CreateAsync(user);
        }

        public async Task<User> GetUserAsync(int id)
        {
            return await _baseRepo.GetAsync(id);
        }

        public async Task<IEnumerable<User>> GetUsersAsync()
        {
            return await _baseRepo.GetAsync();
        }
    }
}

File: src/TinRoll.Data/Repository/UserRepository.fs
namespace TinRoll.Data.Repository

open TinRoll.Data

type IUserRepository =
    interface
        abstract member GetUserAsync : int -> User option Async
        abstract member GetUsersAsync : unit -> User list Async
        abstract member CreateUserAsync : User -> int Async
    end


type UserRepository(context: TinRollContext) =
    interface IUserRepository with
        member this.GetUserAsync id =
            async {
                return context.Users
                       |> Seq.tryFind (fun u -> u.Id = id)
            }

        member this.GetUsersAsync () =
            async {
                return context.Users |> Seq.toList
            }

        member this.CreateUserAsync entity =
            async {
                context.Users.Add(entity) |> ignore
                context.SaveChangesAsync()
                    |> Async.AwaitTask
                    |> Async.RunSynchronously
                    |> ignore
                return entity.Id
            }